octodns/provider/googlecloud.py

```diff
@@ -0,0 +1,343 @@
+        # specified with the GOOGLE_APPLICATION_CREDENTIALS environment
+        # variable. (https://console.cloud.google.com/apis/credentials)
+        #
+        # The project to work on (not required)
+        # project: foobar
+        #
+        # The File with the google credentials (not required). If used, the
+        # "project" parameter needs to be set, else it will fall back to the
+        #  "default credentials"
+        # credentials_file: ~/google_cloud_credentials_file.json
+        #
+    """
+
+    SUPPORTS = set(('A', 'AAAA', 'CAA', 'CNAME', 'MX', 'NAPTR',
+                    'NS', 'PTR', 'SPF', 'SRV', 'TXT'))
+    SUPPORTS_GEO = False
+
+    def __init__(self, id, project=None, credentials_file=None,
+                 *args, **kwargs):
+
+        if credentials_file:
+            self.gcloud_client = dns.Client.from_service_account_json(
+                credentials_file, project=project)
+        else:
+            self.gcloud_client = dns.Client(project=project)
+
+        # Logger
+        self.log = getLogger('GoogleCloudProvider[{}]'.format(id))
+        self.id = id
+
+        super(GoogleCloudProvider, self).__init__(id, *args, **kwargs)
+
+    def _apply(self, plan):
+        """Required function of manager.py to actually apply a record change.
+
+            :param plan: Contains the zones and changes to be made
+            :type  plan: octodns.provider.base.Plan
+
+            :type return: void
+        """
+        desired = plan.desired
+        changes = plan.changes
+
+        self.log.debug('_apply: zone=%s, len(changes)=%d', desired.name,
+                       len(changes))
+
+        # Get gcloud zone, or create one if none existed before.
+        gcloud_zone = self._get_gcloud_zone(desired.name, create=True)
+
+        gcloud_changes = gcloud_zone.changes()
+
+        for change in changes:
+            class_name = change.__class__.__name__
+            _rrset_func = getattr(
+                self, '_rrset_for_{}'.format(change.record._type))
+
+            if class_name in 'Create':
+                gcloud_changes.add_record_set(
+                    _rrset_func(gcloud_zone, change.record))
+            elif class_name == 'Delete':
+                gcloud_changes.delete_record_set(
+                    _rrset_func(gcloud_zone, change.record))
+            elif class_name == 'Update':
+                gcloud_changes.delete_record_set(
+                    _rrset_func(gcloud_zone, change.existing))
+                gcloud_changes.add_record_set(
+                    _rrset_func(gcloud_zone, change.new))
+            else:
+                raise RuntimeError('Change type "{}" for change "{!s}" '
+                                   'is none of "Create", "Delete" or "Update'
+                                   .format(class_name, change))
+
+        gcloud_changes.create()
+        i = 1
+        while gcloud_changes.status != 'done':
```

Comment:
Cool that gcloud has a transactional API. Nothing to do with the provider, but do you know if it's implemented in an atomic/all or nothing matter?

This `while` loop could result in an infinite loop if there's any sort of error state since it would never hit the done state. It'd generally be safer to flip things around and do a `while gcloud_changes.status == 'whatever-the-in-progress-state-is':`

That said we'll want to eventually give up if we don't change state (something really went wrong) so I'd probably suggest reworking this to look something more like:

```python
for in in range(60):
    gcloud_changes.reload()
    if gcloud_changes.status != 'whatever-the-in-progress-state-is':
        # we're no longer in progress so break the loop
        break
    # better name, but nice to make this a constant so you can override it when 
    # testing this case to avoid delaying in a unit test
    sleep(self.CHANGE_LOOP_WAIT)
if gcloud_changes.status != 'done':
    # this is an error we should raise
# if we're here we're done
````

Reply:
Yes these changes are Atomic, and if one fails it'll apply nothing. I was was thinking the very same thing that you'd probably want some sort of eventual timeout. I will add something similar to what you are suggesting.